CIP-XXX/CIP-XXX.md

```diff
@@ -0,0 +1,1408 @@
+| $7$       | Clear         |
+
+Put another way, we can view $w[i] = 1$ to mean that the $(i + 1)$ th least significant
+digit in $w$'s binary representation is $1$, and likewise, $w[i] = 0$ would mean
+that the $i$th least significant digit in $w$'s binary representation is $0$.
+Continuing with the above example, $42$ is represented in binary as `00101010`;
+we can see that the second-least-significant, fourth-least-significant, and
+sixth-least-significant digits are `1`, and all the others are zero. This
+description mirrors the way bytes are represented on machine architectures.
+
+We now extend the above scheme to `BuiltinByteString`s. Let $b$ be a
+`BuiltinByteString` whose length is $n$, and let $i \in 0, 1, \ldots, 8 \cdot n - 1$. 
+For any $j \in 0, 1, \ldots, n - 1$, let $j^{\prime} = n - j - 1$. We say that the bit 
+at $i$ in $b$ is set if
+
+$$
+b\left\\{\left(\left\lfloor \frac{i}{8} \right\rfloor\right)^{\prime}\right\\}[i\mod 8] = 1
+$$
+
+We define the bit at $i$ in $b$ being clear analogously. Similarly to bits in a
+byte, we define $b[i]$ to be $1$ when the bit at $i$ in $b$ is set, and $0$
+otherwise; similarly to bytes, we term this the _value_ at index $i$ in $b$.
+
+As an example, consider the `BuiltinByteString` `[42, 57, 133]`: that is, the
+`BuiltinByteString` $b$ such that $b\\{0\\} = 42$, $b\\{1\\} = 57$ and $b\\{2\\}
+= 133$. We observe that the range of 'valid' bit indexes $i$ into $b$ is in 
+$[0, 3 \cdot 8 - 1 = 23]$. Consider $i = 4$; by the definition above, this
+corresponds to the _byte_ index 2, as $\left\lfloor\frac{4}{8}\right\rfloor =
+0$, and $3 - 0 - 1 = 2$ (as $b$ has length $3$). Within the byte $133$, this
+means we have $\left\lfloor\frac{133}{2^4}\right\rfloor \mod 2 \equiv 0$. Thus,
+$b[4] = 0$. Consider instead the index $i = 19$; by the definition above, this
+corresponds to the _byte_ index 0, as $\left\lfloor\frac{19}{8}\right\rfloor =
+2$, and $3 - 2 - 1 = 0$. Within the byte $42$, this means we have
+$\left\lfloor\frac{42}{2^3}\right\rfloor\mod 2 \equiv 1$. Thus, $b[19] = 1$. 
+
+Put another way, our _byte_ indexes run 'the opposite way' to our _bit_ indexes.
+Thus, for any `BuiltinByteString` of length $n$, we have _bit_ indexes relative
+_byte_ indexes as follows:
+
+```
+| Byte index | 0                              | 1  | ... | n - 1                         |
+|------------|--------------------------------|----| ... |-------------------------------|
+| Byte       | w0                             | w1 | ... | w(n - 1)                      |
+|------------|--------------------------------|----| ... |-------------------------------|
+| Bit index  | 8n - 1 | 8n - 2 | ... | 8n - 8 |   ...    | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
+```
+
+### Operation semantics
+
+We describe precisely the operations we intend to implement, and their
+semantics. These operations will have the following signatures:
+
+* `builtinLogicalAnd :: BuiltinBool -> BuiltinByteString -> BuiltinByteString ->
```

Comment:
The names are odd. We don't normally prefix all operations with "builtin", we do sometimes indicate something about the type. So if anything I would have thought we might want "bytestringLogicalAnd" or something.

Reply:
I'm open to different names. I ended up going with `bitwiseLogicalAnd` in the PR to Plutus Core.

Reply:
I've renamed the operations as follows:

* `builtinLogicalAnd` -> `bitwiseLogicalAnd` (similar for OR, XOR and complement)
* `builtinReadBit` -> `readBit`
* `builtinSetBits` -> `writeBits`
* `builtinReplicate` -> `replicateByteString`

This matches the choices I made when defining these operations in the [Plutus PR](https://github.com/IntersectMBO/plutus/pull/5970) designed to implement this CIP.